Objects/listobject.c

```diff
@@ -2682,16 +2685,34 @@ list_index_impl(PyListObject *self, PyObject *value, Py_ssize_t start,
         if (stop < 0)
             stop = 0;
     }
-    for (i = start; i < stop && i < Py_SIZE(self); i++) {
-        PyObject *obj = self->ob_item[i];
-        Py_INCREF(obj);
-        int cmp = PyObject_RichCompareBool(obj, value, Py_EQ);
-        Py_DECREF(obj);
-        if (cmp > 0)
-            return PyLong_FromSsize_t(i);
-        else if (cmp < 0)
-            return NULL;
-    }
+    if (!key)
+        for (i = start; i < stop && i < Py_SIZE(self); i++) {
+            PyObject *obj = self->ob_item[i];
+            /* take reference to item, in case comparison methods modify list */
+            Py_INCREF(obj);
+            int cmp = PyObject_RichCompareBool(obj, value, Py_EQ);
+            Py_DECREF(obj);
+            if (cmp > 0)
+                return PyLong_FromSsize_t(i);
+            else if (cmp < 0)
+                return NULL;
+        }
+    else
+        for (i = start; i < stop && i < Py_SIZE(self); i++) {
+            PyObject *item = self->ob_item[i];
+            /* take reference to item, in case key modifies list */
+            Py_INCREF(item);
+            PyObject *obj = PyObject_CallFunctionObjArgs(key, item, NULL);
+            Py_DECREF(item);
+            if (!obj)
+                return NULL;
+            int cmp = PyObject_RichCompareBool(obj, value, Py_EQ);
+            Py_DECREF(obj);
+            if (cmp > 0)
+                return PyLong_FromSsize_t(i);
+            else if (cmp < 0)
+                return NULL;
```

Comment:
```suggestion
            if (cmp < 0) {
                return NULL;
            }
            return PyLong_FromSsize_t(i);
```

Reply:
it is not eq when cmp == 0

Reply:
Yep, you're right.
There's should be 
```c
if (cmp == -1) {
    return NULL;
}
else if (cmp == 1 ) {
    return PyLong_FromSsize_t(i);
}

Reply:
This is the original code, only indented.  See newest version.